### DISCUSSION TITLE
Hide some specific branches from `git log` (along with my dirty implementation)

### DISCUSSION THREAD

**Discussion by kuanyui** [magit/magit#5211] [Show and tell]
# Origin
There are some real-world large-scale projects have lots of messy and complex branches for different customers, but it's very difficult to read in Magit when I only want to focus on the branches I need in `M-x magit-log-all-branches` and `M-x magit-log-all`. 

For example, need to cherry-pick across branches, so use `magit-log-all-branches`, but want to exclude some unneeded branches.

The git built-in option `log.excludeDecoration` in `.git/config` works, but it is merely able to hide **the display of `branch_name`, `tag_name`**, while **the tree graph is still rendered in log list.**

So I customized a new option `magitExcludeBranch` to add into `.git/config`, only works for Magit's log list, to solve this problem.

> I've been using this hack for 5 months and feels useful, so post here for share (perhaps others also need?) or as the idea of a potential new feature of Magit. 

# Implementations
I overwrote the two functions in `magit`: `magit-log-all-branches` and `magit-log-all`:
```elisp
  ;; ======================================================
  ;; Exclude some custom branches when magit-log
  ;; ======================================================
  (defun my-magit-get-exclude-branch-from-gitconfig ()
    "Please specify the patterns of branches which you want to exclude in .git/config:
[log]
	magitExcludeBranch = \"master\"
	magitExcludeBranch = \"foo*\"
  "
    (let* ((git-config (plist-get (my-git-get-repo-paths ".") :git-config))
	   (branches '()))
      (with-temp-buffer
	(insert-file-contents git-config)
	(goto-char (point-min))
	(when (re-search-forward "\\[log\\]" nil t)
	  (while (re-search-forward "^\\([[:space:]]+\\)magitExcludeBranch *= \\(.+\\)" nil 'no-error)
	    (let ((branch (match-string 2)))
	      (if (or (string-match "^\".+\"$" branch)
		      (string-match "^'.+'$" branch))
		  (push (substring branch 1 -1) branches)
		(push branch branches))
	      ))))
      branches
      ))

  (defun my-magit-get-exclude-options (type)
    "
   When `git log`:

   1. For --branches + --remotes:

       --exclude='foo' --exclude='master*' --branches
       --exclude='*/foo' --exclude='*/master*' --exclude='*/HEAD' --remotes

   2. If use `--all`, we only need:

       --exclude='refs/tags/*' --exclude='refs/remotes/origin/HEAD' --exclude='**/foo' --exclude='**/master*' --all
  "
    (mapcar (lambda (patt) (concat "--exclude=" patt))
	    (cond ((eq type 'all)
		   `("refs/tags/*"
		     "refs/remotes/origin/HEAD"
		     ,@(mapcar (lambda (patt) (format "**/%s" patt)) (my-magit-get-exclude-branch-from-gitconfig))))
		  ((eq type 'branches)
		   (my-magit-get-exclude-branch-from-gitconfig))
		  ((eq type 'remotes)
		   `("*/HEAD"
		     ,@(mapcar (lambda (patt) (format "*/%s" patt)) (my-magit-get-exclude-branch-from-gitconfig))))
		  )))

  ;; Overwrite magit's function
  (defun magit-log-all-branches (&optional args files)
    "Show log for all local and remote branches and `HEAD'."
    (interactive (magit-log-arguments))
    (magit-log-setup-buffer (if (magit-get-current-branch)
				`(,@(my-magit-get-exclude-options 'branches) "--branches" ,@(my-magit-get-exclude-options 'remotes) "--remotes")
			      `("HEAD" ,@(my-magit-get-exclude-options 'branches) "--branches" ,@(my-magit-get-exclude-options 'remotes) "--remotes"))
			    args files))

  ;; Overwrite magit's function
  (defun magit-log-all (&optional args files)
    "Show log for all references and `HEAD'."
    (interactive (magit-log-arguments))
    (magit-log-setup-buffer (if (magit-get-current-branch)
				`(,@(my-magit-get-exclude-options 'all) "--all")
                              `("HEAD" ,@(my-magit-get-exclude-options 'all) "--all"))
			    args files))
```

# Usage
In the git repo's `.git/config`, add field(s) `magitExcludeBranch` in `[log]`
```conf
[log]
	excludeDecoration = *customerA*
	excludeDecoration = refs/remotes/origin/customerA*
	excludeDecoration = refs/heads/customeAr*
	magitExcludeBranch = "customerA*"
	magitExcludeBranch = "anotherCustomer*"
```

And the branches whose name starts with `customerA` will be totally omitted from the rendering of magit log.

**Comment by tarsius** [comment]
Are you aware that you can add additional *negative* refspecs to `remote.<name>.fetch`? See `git-fetch(1)` for details.

For example:

```
[remote "origin"]
	url = tarsius@git.savannah.gnu.org:/srv/git/emacs.git
	fetch = +refs/heads/*:refs/remotes/origin/*
	fetch = ^refs/heads/old-branches/*
	fetch = ^refs/heads/other-branches/*
        [... and so on. There are A LOT of garbage refs in this repository! ]
```

This won't delete the remote tracking branches that have already been fetched.

You can easily delete them in magit. `y`, mark multiple branches using the region, `k`, say yes to the first two questions, but make sure you say NO when asked whether to also delete on the remote.

**Reply by kuanyui** [reply]
> Are you aware that you can add additional *negative* refspecs to `remote.<name>.fetch`? See `git-fetch(1)` for details.
> 
> For example:
> 
> ```
> [remote "origin"]
> 	url = tarsius@git.savannah.gnu.org:/srv/git/emacs.git
> 	fetch = +refs/heads/*:refs/remotes/origin/*
> 	fetch = ^refs/heads/old-branches/*
> 	fetch = ^refs/heads/other-branches/*
>         [... and so on. There are A LOT of garbage refs in this repository! ]
> ```
> 
> This won't delete the remote tracking branches that have already been fetched.
> 

Yes, I knew this before, but inconvenient for me. 

My solution can hide specific branch temporarily & permanently, also can switch between hide/show  immediately according to different requirements, without fetching nor setting args/options via transition.el every time when opening the git log list, which at least fits my requirements more.


> You can easily delete them in magit. `y`, mark multiple branches using the region, `k`, say yes to the first two questions, but make sure you say NO when asked whether to also delete on the remote.

I have to say that I frequently mistakenly do some wrong operations in Magit because I'm too lazy and inpatient to read the questions appeared in minibuffer every time, so less questions is definitely better for me.

**Reply by tarsius** [reply]
Instead of the first function, use `(magit-get-all "log.magitExcludeBranch")`.

Instead of redefining `magit-log-all[-branches]`, it would be better to use an `:override` advice. A cleaner implementation strategy *might* be to *instead* advice `magit-log-arguments`.

**Comment by kuanyui** [comment]
> Instead of the first function, use `(magit-get-all "log.magitExcludeBranch")`.

Sorry, I missed to post the definition of the function `my-git-get-repo-paths` used in the first post... Maybe it can be done via `(magit-get-all "log.magitExcludeBranch")` directly, I admit that I have no more motivation to refactor it as long as it works...

```elisp
  ;; ======================================================
  ;; toggle excludeDecoration
  ;; ======================================================
  (defun my-git-cmd (fmt &rest args)
    (let ((raw (shell-command-to-string (apply #'format fmt args))))
      (if (string-suffix-p "\n" raw)
	  (setq raw (substring raw 0 -1)))
      raw))

  (defun my-git-get-repo-paths (from-path)
    "Returns a plist"
    (let* ((default-directory (if from-path (expand-file-name from-path) default-directory))
	   (nearest-git-parent (file-truename (locate-dominating-file from-path ".git")))  ;; ex: `~/.emacs.d/`
	   (is-regular (file-directory-p (file-name-concat nearest-git-parent ".git")))
	   ;; priority 1; submodule; superproject-nearest-git-parent
	   (raw-submodule (my-git-cmd "git rev-parse --show-superproject-working-tree"))
	   ;; priority 2, regular (oneline) + worktree (oneline) + submodule (multilines);
	   (raw-commondir (my-git-cmd "git rev-parse --show-superproject-working-tree --path-format=absolute --git-common-dir"))
	   (git-config-abspath (expand-file-name (my-git-cmd "git rev-parse --git-path config")))
	   )
      (cond ((null git-config-abspath) nil)
	    (is-regular
	     (list :type 'regular
		   :superproject-workingdir nearest-git-parent
		   :project-workingdir nearest-git-parent
		   :git-config git-config-abspath
		   ))
	    ((> (length raw-submodule) 0)
	     (list :type 'submodule
		   :superproject-workingdir raw-submodule
		   :project-workingdir nearest-git-parent
		   :git-config git-config-abspath
		   )
	     )
	    (t
	     (list :type 'worktree
		   :superproject-workingdir raw-commondir
		   :project-workingdir nearest-git-parent
		   :git-config git-config-abspath
		   )
	     ))))
```

**Comment by tarsius** [comment]
Thanks for implementing this. Maybe someone else will find this useful, so I am moving it to the "Show and tell" discussion category[^1], but I do not intend to add something like this to magit.

[^1]: Or was it there already? When I just scrolled up to put it in there, that was already there all along, or maybe I just put it there a minute ago? :stuck_out_tongue_winking_eye:
